src/lib/runner.rs

```diff
@@ -122,18 +122,37 @@ pub fn generate(missile: &Missile, launch_parameters: &LaunchParameter, timestep
 
 
 		let engine_stage;
-
-		if (f64::from(i) * timestep) < missile.timefire0 {
-			a = ((missile.force0 - drag_force) / missile.mass) - gravity;
-			engine_stage = "0";
-		} else if missile.timefire1 != 0.0 && (f64::from(i) * timestep) < missile.timefire1 {
-			a = ((missile.force1 - drag_force) / missile.mass) - gravity;
-			engine_stage = "1";
-		} else {
-			a = (-drag_force / missile.mass_end) - gravity;
-			engine_stage = "-";
+		let mass;
+		let force;
+
+		let burn_0 = 0.0..missile.timefire0;
+		let burn_1 = burn_0.end..missile.timefire1;
+
+		let burn_time = f64::from(i) * timestep;
+		let compute_mass = |mass_end, timefire| {
+			missile.mass + (missile.mass - mass_end) * (burn_time/ timefire)
+		};
+		match () {
+			_ if burn_0.contains(&burn_time) => {
+				mass = compute_mass(missile.mass_end, missile.timefire0);
+				force = missile.force0;
+				engine_stage = "0";
+			}
+			_ if burn_1.contains(&burn_time) => {
+				mass = compute_mass(missile.mass_end1, missile.timefire1);
+				force = missile.force1;
+				engine_stage = "1";
+			}
+			_ => {
+				mass = missile.mass_end;
```

Comment:
if (**missile.mass_end1** exists _(!= 0)?_)
then
mass = missile.mass_end1;
else 
mass = missile.mass_end;

or  you will actually make the missile heavier at end of the burn